Iterables (#133)
javascript-tutorial/fa.javascript.info: 1-js/05-data-types/06-iterable/article.md

## ویژگی Symbol.iterator

ما می‌توانیم به راحتی مفهوم حلقه‌پذیرها را با ایجاد حلقه‌پذیر خودمان درک کنیم.

برای مثال، ما یک شیء داریم که آرایه نیست، اما برای استفاده در `for..of` مناسب است.

مانند یک شیء `range` که بازه‌ای از اعداد را نشان می‌دهد:

```js
let range = {
  from: 1,
  to: 5
};

// :کار کند for..of می‌خواهیم که
// for(let num of range) ... num=1,2,3,4,5
```

برای اینکه شیء `range` را حلقه‌پذیر کنیم (و به این ترتیب بگذاریم `for..of` کار کند) ما نیاز داریم که یک متد به اسم `Symbol.iterator` را به شیء اضافه کنیم ( یک سمبل خاص درون ساخت که فقط برای این کار است).

1. زمانی که `for..of` شروع می‌شود، متد را یک بار صدا می‌زند (یا اگر پیدا نشود ارور می‌دهد). متد باید یک *حلقه‌زننده* را برگرداند -- شیءای که متد `next` را دارد.
2. همینطور رو به جلو، `for..of` *تنها با شیء برگردانده شده* کار می‌کند.
3. زمانی که `for..of` مقدار بعدی را نیاز دارد، روی آن شیء `next()` را صدا می‌زند.
4. نتیجه `next()` باید به شکل `{done: Boolean, value: any}` باشد که `done=true` به معنی پایان حلقه‌زدن است، در غیر این صورت `value` مقدار بعدی خواهد بود.

> اینجا پیاده‌سازی کامل را برای `range` به همراه اظهارات داریم:

Reviewer: ```suggestion
اینجا پیاده‌سازی کامل `range` را به همراه ملاحظات داریم:
```

```js run
let range = {
  from: 1,
  to: 5
};

> // 1. ابتدا این متد صدا زده می‌شود for..of با استفاده از

Reviewer: ```suggestion
// 1. در ابتدا این متد صدا زده می‌شود for..of با صدازدن
```

اتفاق مشابهی برای حلقه‌پذیر می‌افتد:

```js
// از مثال بالا گرفته شده است range فرض می‌کنیم که
let arr = Array.from(range);
alert(arr); // 1,2,3,4,5 (آرایه کار می‌کند toString تبدیل)
```

سینتکس کامل برای `Array.from` به اجازه فراهم کردن یک تابع «طراحی» هم می‌دهد:
```js
Array.from(obj[, mapFn, thisArg])
```

آرگومان اختیاری دوم `mapFn` می‌تواند تابعی باشد که روی تمام المان‌ها قبل از اینکه به آرایه اضافه شوند اعمال می‌شود و `thisArg` اجازه می‌دهد که برای آن `this` قرار دهیم.

برای مثال:

```js
// از مثال بالا گرفته شده است range فرض می‌کنیم

// به توان 2 رساندن هر عدد
let arr = Array.from(range, num => num * num);

alert(arr); // 1,4,9,16,25
```

اینجا ما از `Array.from` برای تبدیل یک رشته به آرایه‌ای از کاراکترها استفاده می‌کنیم:

```js run
let str = '𝒳😂';

// به آرایه‌ای از کاراکترها str تقسیم
let chars = Array.from(str);

alert(chars[0]); // 𝒳
alert(chars[1]); // 😂
alert(chars.length); // 2
```

برخلاف `str.split`، این روش بر اساس طبیعت حلقه‌پذیری رشته کار می‌کند و به همین دلیل، درست مانند `for..of`، با جفت‌های جایگیر به درستی کار می‌کند.

> از لحاظ فنی اینجا هم کار مشابهی را انجام می‌دهد:

Reviewer: ```suggestion
از لحاظ فنی این هم کار مشابهی را انجام می‌دهد:
```

```js run
let str = '𝒳😂';

let chars = []; // هم از درون این حلقه را اجرا می‌کند Array.from
for (let char of str) {
  chars.push(char);
}

alert(chars);
```

> ...اما این کوتاه‌تر است.

Reviewer: ```suggestion
...اما این روش کوتاه‌تر است.
```
